### PR TITLE
deploymentstatus: move g_versionbitscache global to ChainstateManager

### PR DESCRIPTION
**Pull request by ajtowns**
Gives `ChainstateManager` a reference to the `CChainParams` its working on, and simplifies some of the functions that would otherwise take that as a parameter. Removes the `g_versionbitscache` global by moving it into `ChainstateManager`.

### CODE REVIEW

**Comment by dongcarl** [comment]
Concept ACK!

Also wanted to note that in one of my branches for `libbitcoinkernel` I give `ChainstateManager` a `CChainParams` as well, so it definitely makes sense!

**Comment by ajtowns** [comment]
Should be ready for review now

**Comment by ajtowns** [comment]
Rebased on top of #22564 also no longer introduce a circular dependency between deploymentstatus and validation.

File: src/validation.h
Comment (dongcarl) [inline src/validation.h:942]: Could you explain a bit why `m_versionbitscache` is marked `mutable`?

File: src/validation.h
Comment (dongcarl) [inline src/validation.h:858]: Perhaps these additions could be made in the previous commit where we introduce the private `m_chainparams` member

File: src/validation.h
Comment (dongcarl) [inline src/validation.h:494]: By this, do you mean all references to `(CChainState::)m_params` by `CChainState` member functions should instead be references to `m_chainman.GetParams()`?

File: src/validation.cpp
Comment (dongcarl) [inline src/validation.cpp:3365]: nit: I'm not sure that this change is 100% necessary. The change seems to trade one reference for another, and since `ChainstateManager` is much larger in scope, it is harder to see after the change that this is a function that is limited in scope.

However, this is not at all a bit deal and up to subjective taste so feel free to ignore!

File: src/validation.cpp
Comment (dongcarl) [inline src/validation.cpp:1897]: Perhaps we can use a `const VersionBitsCache&` here?

**Review by dongcarl — COMMENTED** [review]
Medium Code-Review ACK 555eb9344ca83f772fa810570e1aa07881d33a41

💥 🔫  Another mutable global bites the dust!

File: src/validation.h
Comment (ajtowns) [reply src/validation.h:942]: It's a cache, so its existence doesn't impact the logical API of the class -- if we didn't use a cache, you could have a `const ChainstateManager` and get answers as to whether a deployment is active/enabled and what version to use for new blocks; so adding a cache shouldn't mean that you can only do that with a non-const `ChainstateManager`.

ie, I'm treating `const ChainstateManager` as being "logically" constant, not bitwise constant.

File: src/validation.h
Comment (ajtowns) [reply src/validation.h:494]: Yeah. Or perhaps add a `GetParams()` and `GetConsensus()` to `CChainState` that just call `m_chainman.GetParams()` or `m_chainman.GetConsensus()`.

File: src/validation.cpp
Comment (ajtowns) [reply src/validation.cpp:3365]: It needs to call `DeploymentActive(.., SEGWIT)`, and that's something that the `ChainstateManager` is responsible for being able to answer. At least that was my logic.

Note that it's a `const` member function, so it's not doing anything to change the chain state, if that counts as reducing the scope?

File: src/validation.cpp
Comment (ajtowns) [reply src/validation.cpp:1897]: Could, but it should all get optimised away since `WarningBitsConditionChecker` is only briefly instantiated while it's being checked.

File: src/validation.h
Comment (ajtowns) [reply src/validation.h:494]: (I didn't start on this here since I imagine it would have a bunch of conflicts with other PRs)

File: src/validation.h
Comment (dongcarl) [reply src/validation.h:942]: Sounds good!

File: src/validation.h
Comment (dongcarl) [reply src/validation.h:494]: Sounds good!

File: src/validation.cpp
Comment (dongcarl) [reply src/validation.cpp:3365]: Sounds good!

File: src/validation.cpp
Comment (dongcarl) [reply src/validation.cpp:1897]: Sure!

**Comment by MarcoFalke** [comment]
Are you sure this is ready for review? Looks like it is based on #22564 for now, which might change its commits. So maybe leave it in draft until #22564 is merged? (Reviewers can still ACK your commits here, but I don't think it can be merged before #22564).

Also, I am wondering what to do with #24917. Should that go before/after #22564 or should it be included there? cc @dongcarl @ajtowns

**Comment by ajtowns** [comment]
> Are you sure this is ready for review? Looks like it is based on #22564 for now,

Yeah, there's a trivial conflict between #22564 and these patches in that they both change the call to `g_versionbitscache.Clear()`; seemed more sensible to review the changes in #22564 first, so I rebased the changes here on top of that.

> Also, I am wondering what to do with #24917. Should that go before/after #22564 or should it be included there?

Seems fine to keep it separate, and 22564 seems higher priority and not blocked on anything; so I'd say 22564 first and keep 24917 separate.

File: src/validation.h
Comment (MarcoFalke) [inline src/validation.h:856]: 4cbc4e94b764ba2b4a79610fadbb092a7bdd076e: What about a public `m_params` and `m_consensus`? There shouldn't be any risk associated with read-only references. And a public member removes the need for a function wrapper that only returns the member.

File: src/validation.h
Comment (MarcoFalke) [inline src/validation.h:18]: 0fc617a406232e21bdc8b39b2f8308f965775825 Any reason to add this include when the existing fwd-decl is enough?

**Review by maflcko — APPROVED** [review]
ACK

File: src/validation.h
Comment (ajtowns) [reply src/validation.h:856]: 6 to one, half a dozen the other? I thought accessors were more common than extra variables, and make it a bit clearer that they can't get out of sync? *shrug*

File: src/validation.h
Comment (ajtowns) [reply src/validation.h:18]: Adding `#include <chainparams.h>` in the previous commit already brings `consensus/params.h` in so the fwd declaration isn't saving anything. Later, `DeploymentActive*` use info from `consensus/params.h` too.

**Comment by dongcarl** [comment]
ACK d490648b2591f523a096f926e3f29e34d72da55a

-----

Since last review @ 555eb9344ca83f772fa810570e1aa07881d33a41
- Rebased over merge of #22564
- Addressed https://github.com/bitcoin/bitcoin/pull/24595#discussion_r852322280

File: src/validation.h
Comment (MarcoFalke) [reply src/validation.h:18]: I think for passing references or pointers, a fwd-decl is enough.

iwyu seems to agree:


```
validation.h should remove these lines:
- #include <config/bitcoin-config.h>  // lines 10-10
- #include <consensus/params.h>  // lines 18-18
- #include <stdint.h>  // lines 39-39
- #include <util/hasher.h>  // lines 30-30
- class CBlockTreeDB;  // lines 46-46
- class CTxMemPool;  // lines 47-47
- struct AssumeutxoData;  // lines 53-53
- struct ChainTxData;  // lines 49-49
- struct DisconnectedBlockTransactions;  // lines 50-50
- struct LockPoints;  // lines 52-52

Comment (ajtowns) [reply src/validation.h:18]: Ah, iwyu treats the `DeploymentActive` stuff as recommending that deploymentstatus.h add the `#include <consensus/params.h>` not validation.h. I've got some later commits that move those functions into validation.h which updates iwyu's recommendations to a #include.

Since it's already included indirectly by chainparams.h I don't really see the include as making it worse, but it's not really better either, so will change it back if there are other fixes to bundle it with.

**Comment by ajtowns** [comment]
Rebased past #24538

File: src/validation.h
Comment (ajtowns) [reply src/validation.h]: Done now

**Comment by dongcarl** [comment]
reACK bb5c24b120a3ac7df367a1c5d9b075ca564efb5f

File: src/validation.cpp
Comment (MarcoFalke) [inline src/validation.cpp:1458]: 5c67e84d37d452e9186a6357e5405fabeff241c7 nit: `.m_chainman.GetParams()`

Seems odd to add a TODO and violate it in the same commit (     /* TODO: replace with m_chainman.GetParams() */)

File: src/test/versionbits_tests.cpp
Comment (MarcoFalke) [inline src/test/versionbits_tests.cpp]: eca22c726ac48b4216bb68cc0f0bbd655c43ac12 nit: Later diff could be smaller by a few lines if you added a `VersionBitsCache& vbcache{g_versionbitscache};` in this commit already.

**Review by maflcko — APPROVED** [review]
review ACK bb5c24b120a3ac7df367a1c5d9b075ca564efb5f 📙

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK bb5c24b120a3ac7df367a1c5d9b075ca564efb5f 📙
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUi6+Qv8Dw67I3SBVnRjV/M10aNNkHAduz39CLI+zz2TwT/0IyAbgOEB9QAwzWS1
RFEQdpklN4Vxu+4qwkpEUWtl0cpJV8buVab33eUgDI1WuN/jo8nzF0GWpFwkHUJp
4VEYDP+9zWssKUGgtsV0LPadXFsk74f9KW0q1FN+UljnEBBksPtRfGok1TQSzesH
m0JlsdZcOBMFjNA+k7cQ3xyZi4b7tERRe8NZ1dewvHMBGpeh5J/n/TwHVO/4NO4Z
S7qAENfXCP8PWblW5LgvUSlxpyKbwma5ygHe7M1/RnmzMQO1BRkcq0x78ktyxWwU
Nq1hh4/ETvJgy3j3FIIm2AZeRRo+I1HhhJ1DuqKsqs3KNsM/+jToHrkel7Qi4WtT
nzyK66L2Tb/MKGsXq2lyU9UC2U4REW3FTi0btmEnRl35RCUyhWbNppl2eZDaheHk
NKnJV/wgZvgJUVra4Gt6pverD2dPL9/PA6V/M27tF1R9/buUMeEyQYVcBPBw5zSB
ySIzD/fI
=er/p
-----END PGP SIGNATURE-----
```


</details>

File: src/test/versionbits_tests.cpp
Comment (MarcoFalke) [inline src/test/versionbits_tests.cpp:416]: Removed chainman completely from this test in https://github.com/bitcoin/bitcoin/pull/25125

**Pull request merged**
